src/cargo/core/registry.rs

```diff
@@ -892,12 +927,30 @@ fn lock(
             }
         }
 
-        // If this dependency did not have a locked version, then we query
-        // all known locked packages to see if they match this dependency.
-        // If anything does then we lock it to that and move on.
-        let v = locked
-            .get(&(dep.source_id(), dep.package_name()))
-            .and_then(|vec| vec.iter().find(|&&(id, _)| dep.matches_id(id)));
+        // If this dependency did not have a locked version, we check if multiple
+        // versions of the dependency already exists in the lockfile. In this case we want
+        // to do the full resolving instead of locking a specific version.
+        // Otherwise, we query all known locked packages to see if they match this dependency,
+        // and if anything does then we lock it to that and move on.
+
```

Comment:
My instinct has long been that this heuristic should just be removed. Apparently I tried that when I was a little baby contributor new to open source in #5718. A lot has changed since then, so it might work now.

Another approach discussed in that PR is to remove this heuristic and double our calls to `resolve`. We try resolving with `--offline` and only try normal resolution after the `--offline` did not work. But now that I say that (I think, without having try it), that would lead to `cargo update` randomly updating unrelated GIT dependencies. In addition to the performance impacts.

Reply:
After some more investigation, this heuristic seems to be loadbearing for cleaning up different ways to refer to the same git package. For example a dependency might require the default branch (or the `master` branch) but what's available in the lock file is a particular checkout. This heuristic identifies that the particular hash could match and uses it. This seems accidental to me. I think there must be a better way of handling this. If we had a `matches_with_compatible_git_source` when iterating over `locked_deps` things could be solved much more elegantly. In the time I had available I did not find one laying around, but this still seems a promising direction.

Reply:
This test checks this heuristic and is not related to git dependencies:
https://github.com/rust-lang/cargo/blob/43e3fa90257d899c7281d3808d37cbf569a21587/tests/testsuite/registry.rs#L2487

Reply:
You are correct. I forgot to mention that case. That test was added at the same time as the heuristic. Based on the PR that added it was to allow use of cargo with fewer network requests. This was a big issue at the time because we did not yet have `--offline`. So if cargo decided that you needed a network request and you did not have network access you were stuck. I have not formally discussed it with the team, but my vote would be that we just remove that test.